### ISSUE TITLE
Apparmor  >  swtpm: Could not open UnixIO socket: Permission denied

### ISSUE THREAD

**Issue by TuxVinyards** [quickemu-project/quickemu#487]
#### Expected behaviour

Ubuntu 22.04 MATE

UnixIO socket  works only when I disable Apparmor

#### Actual behaviour

Throws these errors: 

swtpm: Could not open UnixIO socket: Permission denied 

audit: type=1400 audit(1655107403.313:54): apparmor="DENIED" operation="mknod" profile="swtpm" name="/media/xxx/Storage/VMs/windows-11/windows-11.swtpm-sock" pid=3086 comm="swtpm" requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000

Can't copy/paste to/from guest, see Public folder etc

#### Steps to reproduce the behaviour

??  Not sure how to advise on this one. It's permanent for me.

? Is it related to this:  https://github.com/stefanberger/swtpm/issues/297

#### Quickemu output

`
    BOOT: EFI (Windows), OVMF (/usr/share/OVMF/OVMF_CODE_4M.fd), SecureBoot (on).
    Disk: windows-11/disk.qcow2 (128G)
    Just created, booting from windows-11/Win11_EnglishInternational_x64v1.iso
    Boot ISO: windows-11/Win11_EnglishInternational_x64v1.iso
    CD-ROM: windows-11/virtio-win.iso
    Display: SDL, qxl-vga, GL (on), VirGL (off)
    ssh: On host: ssh user@localhost -p 22220
    SPICE: On host: spicy --title "windows-11" --port 5930 --spice-shared-dir /home/xxx/Public
    WebDAV: On guest: dav://localhost:9843/
    TPM: windows-11/windows-11.swtpm-sock (21305)
    swtpm: Could not open UnixIO socket: Permission denied
    Process: Starting windows-11.conf as windows-11 (21307)

`

#### Linux Distribution & Kernel

Distributor ID:	Ubuntu
Description:	Ubuntu 22.04 LTS
Release:	22.04
Codename:	jammy

Linux System-X 5.15.0-37-generic #39-Ubuntu 

#### apparmor_status

50 profiles are loaded.
41 profiles are in enforce mode.

  bla  bla bla

   libvirtd
   libvirtd//qemu_bridge_helper
   swtpm
   virt-aa-helper

Wondering about how safe setting swtpm to complain mode might be?

Thoughts?

```

**Comment by TuxVinyards** [comment]
So,  I have traced the problem to there being no AppArmor  prior to 4 months ago & the start of things here:

[https://bugs.launchpad.net/ubuntu/+source/swtpm/+bug/1968335](url)

I am not sure that this is entirely an Ubuntu thing though.  I think there could be a general problem with quickemu using only relative links & also putting everything in its own custom folder.  Exacerbated on my computer because I have my VM's on a separate partition ...

I quote from the launchpad swtpm  'fix'  -  "Since there is no default location for the tpm unix socket I suggest using "/run/swtpm/sock" which follows the systemd.socket unit conventions"

Looking at  the quickemu code, using absolute references, something like this perhaps (?)  might work: 

```

 # Start TPM

  if [ "${tpm}" == "on" ]; then
    local tpm_args=()
    
   #  ADD this absolute path :

    VmDirPath="$(pwd)"
    
    # shellcheck disable=SC2054
       
    tpm_args+=(socket
            --ctrl type=unixio,path="$VmDirPath/${VMDIR}/${VMNAME}.swtpm-sock"
            --terminate
            --tpmstate dir="$VmDirPath/${VMDIR}"
            --tpm2)
    echo "${SWTPM} ${tpm_args[*]} &" >> "$VmDirPath/${VMDIR}/${VMNAME}.sh"
    ${SWTPM} "${tpm_args[@]}" >> "$VmDirPath/${VMDIR}/${VMNAME}.log" &
    echo " - TPM:      $VmDirPath/${VMDIR}/${VMNAME}.swtpm-sock (${!})"
    sleep 0.25
  fi

```

From around  line 850

Although,  perhaps referencing a standard system folder might be better ...


### Is it safe to remove the profile?

Well, probably not best practice, but as one didn't even exist 4 months ago, probably not that bad, with a bit of caution.

For anyone looking  for the commands:

```
sudo aa-complain /usr/bin/swtpm
sudo apparmor_parser -r /etc/apparmor.d/usr.bin.swtpm
```
I had to do a reboot to get things stable.

SysLog  will  start showing UFW (firewall) references after the VM loads.  Make sure that your firewall is on ...

Also see #448

**Comment by flexiondotorg** [comment]
This looks like a regression to a patch that was added to the packaging of swtpm in Ubuntu.

 - https://bugs.launchpad.net/ubuntu/+source/swtpm/+bug/1968335

**Comment by Craig-Macomber** [comment]
I also hit this on Ubuntu 22.04.

I can confirm this fixed the error:
```
sudo apt-get install apparmor-utils
sudo aa-complain /usr/bin/swtpm
sudo apparmor_parser -r /etc/apparmor.d/usr.bin.swtpm
```

And caused Windows 11 to support "Windows Hello".

I did not need to reboot (the host) afterwards, only the guest.

As @TuxVinyards mentioned (thanks for the workaround!), this has some possible security implications.

**Comment by flexiondotorg** [comment]
This is fixed since swtpm 0.6.3-0ubuntu3.1 in Ubuntu 22.04 and newer.

- https://launchpad.net/ubuntu/+source/swtpm/0.6.3-0ubuntu3.1
